src/MomentumHybridTurbElemKernel.C

```diff
@@ -0,0 +1,209 @@
+  alphaNp1_ = metaData.get_field<ScalarFieldType>(
+    stk::topology::NODE_RANK, "adaptivity_parameter");
+  mutij_ = metaData.get_field<GenericFieldType>(
+    stk::topology::NODE_RANK, "tensor_turbulent_viscosity");
+
+  coordinates_ = metaData.get_field<VectorFieldType>(
+    stk::topology::NODE_RANK, solnOpts.get_coordinates_name());
+
+  MasterElement* meSCS =
+    sierra::nalu::MasterElementRepo::get_surface_master_element(
+      AlgTraits::topo_);
+  get_scs_shape_fn_data<AlgTraits>(
+    [&](double* ptr) { meSCS->shape_fcn(ptr); }, v_shape_function_);
+
+  // add master elements
+  dataPreReqs.add_cvfem_surface_me(meSCS);
+
+  // fields
+  dataPreReqs.add_coordinates_field(
+    *coordinates_, AlgTraits::nDim_, CURRENT_COORDINATES);
+  dataPreReqs.add_gathered_nodal_field(*velocityNp1_, AlgTraits::nDim_);
+  dataPreReqs.add_gathered_nodal_field(*densityNp1_, 1);
+  dataPreReqs.add_gathered_nodal_field(*tkeNp1_, 1);
+  dataPreReqs.add_gathered_nodal_field(*alphaNp1_, 1);
+  dataPreReqs.add_gathered_nodal_field(
+    *mutij_, AlgTraits::nDim_, AlgTraits::nDim_);
+
+  // master element data
+  dataPreReqs.add_master_element_call(SCS_AREAV, CURRENT_COORDINATES);
+  if (shiftedGradOp_)
+    dataPreReqs.add_master_element_call(
+      SCS_SHIFTED_GRAD_OP, CURRENT_COORDINATES);
+  else
+    dataPreReqs.add_master_element_call(SCS_GRAD_OP, CURRENT_COORDINATES);
+}
+
+template <typename AlgTraits>
+void
+MomentumHybridTurbElemKernel<AlgTraits>::execute(
+  SharedMemView<DoubleType**>& lhs,
+  SharedMemView<DoubleType*>& rhs,
+  ScratchViews<DoubleType>& scratchViews)
+{
+  DoubleType w_mutijScs[AlgTraits::nDim_ * AlgTraits::nDim_];
+
+  SharedMemView<DoubleType**>& v_uNp1 =
+    scratchViews.get_scratch_view_2D(*velocityNp1_);
+  SharedMemView<DoubleType*>& v_rhoNp1 =
+    scratchViews.get_scratch_view_1D(*densityNp1_);
+  SharedMemView<DoubleType*>& v_tkeNp1 =
+    scratchViews.get_scratch_view_1D(*tkeNp1_);
+  SharedMemView<DoubleType*>& v_alphaNp1 =
+    scratchViews.get_scratch_view_1D(*alphaNp1_);
+  SharedMemView<DoubleType***>& v_mutij =
+    scratchViews.get_scratch_view_3D(*mutij_);
+
+  SharedMemView<DoubleType**>& v_scs_areav =
+    scratchViews.get_me_views(CURRENT_COORDINATES).scs_areav;
+  SharedMemView<DoubleType***>& v_dndx =
+    shiftedGradOp_ ? scratchViews.get_me_views(CURRENT_COORDINATES).dndx_shifted
+                   : scratchViews.get_me_views(CURRENT_COORDINATES).dndx;
+
+  for (int ip = 0; ip < AlgTraits::numScsIp_; ++ip) {
+
+    // left and right nodes for this ip
+    const int il = lrscv_[2 * ip];
+    const int ir = lrscv_[2 * ip + 1];
+
+    // save off some offsets
+    const int ilNdim = il * AlgTraits::nDim_;
+    const int irNdim = ir * AlgTraits::nDim_;
+
+    // zero out vector that prevail over all components
+    for (int i = 0; i < AlgTraits::nDim_; ++i) {
+      const int offset = i * AlgTraits::nDim_;
+      for (int j = 0; j < AlgTraits::nDim_; ++j) {
+        w_mutijScs[offset + j] = 0.0;
```

Comment:
You can probably choose to define this as (ndim,nDim) to avoid offsets - either way is fine.